server/client.go

```diff
@@ -1116,6 +1102,90 @@ func (c *client) flushOutbound() bool {
 	return true
 }
 
+// This is invoked from flushOutbound() for io/timeout error (slow consumer).
+// Returns a boolean to indicate if the connection has been closed or not.
+// Lock is held on entry.
+func (c *client) handleWriteTimeout(written, attempted int64, numChunks int) bool {
+	if tlsConn, ok := c.nc.(*tls.Conn); ok {
+		if !tlsConn.ConnectionState().HandshakeComplete {
+			// Likely a TLSTimeout error instead...
+			c.markConnAsClosed(TLSHandshakeError, true)
+			// Would need to coordinate with tlstimeout()
+			// to avoid double logging, so skip logging
+			// here, and don't report a slow consumer error.
+			return true
+		}
+	} else if c.expectConnect() && !c.flags.isSet(connectReceived) {
+		// Under some conditions, a connection may hit a slow consumer write deadline
+		// before the authorization timeout. If that is the case, then we handle
+		// as slow consumer though we do not increase the counter as that can be
+		// misleading.
+		c.markConnAsClosed(SlowConsumerWriteDeadline, true)
+		return true
+	}
+
+	// Slow consumer here..
+	atomic.AddInt64(&c.srv.slowConsumers, 1)
+	c.Noticef("Slow Consumer Detected: WriteDeadline of %v exceeded with %d chunks of %d total bytes.",
+		c.out.wdl, numChunks, attempted)
+
+	// We always close CLIENT connections, or when nothing was written at all...
+	if c.kind == CLIENT || written == 0 {
+		c.markConnAsClosed(SlowConsumerWriteDeadline, true)
+		return true
+	}
+	return false
+}
+
+// Marks this connection has closed with the given reason.
+// Sets the closeConnection flag and skipFlushOnClose flag if asked.
+// Depending on the kind of connection, the connection will be saved.
+// If a writeLoop has been started, the final flush/close/teardown will
+// be done there, otherwise flush and close of TCP connection is done here in place.
+// Returns true if closed in place, flase otherwise.
+// Lock is held on entry.
+func (c *client) markConnAsClosed(reason ClosedState, skipFlush bool) bool {
+	if c.flags.isSet(closeConnection) {
+		return false
+	}
+	c.flags.set(closeConnection)
+	if skipFlush {
+		c.flags.set(skipFlushOnClose)
+	}
+	// Save off the connection if its a client or leafnode.
+	if c.kind == CLIENT || c.kind == LEAF {
+		if nc := c.nc; nc != nil && c.srv != nil {
+			go c.srv.saveClosedClient(c, nc, reason)
+		}
+	}
+	// If writeLoop exists, let it do the final flush, close and teardown.
+	if c.flags.isSet(writeLoopStarted) {
+		c.out.sg.Broadcast()
+		return false
+	}
+	// Flush (if skipFlushOnClose is not set) and close in place. If flushing,
+	// use a small WriteDeadline.
+	c.flushAndClose(true)
+	return true
+}
+
+// Based on the kind of connection, returns if a CONNECT is expected or not.
+// Lock is held on entry.
+func (c *client) expectConnect() bool {
+	switch c.kind {
+	case CLIENT:
+		return true
```

Comment:
This is not always true. Think telnet example.

Reply:
The problem is that we treat a write timeout differently when the CONNECT has not been received yet. That is, there is a test that make sure that if the client does not send the CONNECT and the write deadline is hit, we do not report has a slow consumer (will be auth timeout instead).

Reply:
If server has auth set then I think it's ok to say client needs to send connect, but when its not sent I think this should return false for client. WDYT?

Reply:
Ok, this was more involved since I had to access server's info's AuthRequired or TLSRequired, so I moved to flags instead and set it in each of the createClient(), createRoute(), etc... with the same logic that was in expectConnect() switch case.

Reply:
TLS for clients does not mean connect is needed, only auth IIRC.